### PR TITLE
[MIG][13.0] product_pricelist_direct_print: Migration to v13.0

### PR DESCRIPTION
**Pull request by sergio-teruel**
cc @Tecnativa TT21646
@carlosdauden @Tardo Review please

### CODE REVIEW

**Comment by legalsylvain** [comment]
@sergio-teruel, Travis tests related to your module are not passing. Could you check ?

**Comment by sergio-teruel** [comment]
Fixed!! Although the module coverage is low...

**Comment by legalsylvain** [comment]
You can improve it, checking the the result of the call of ``message_composer_action`` and ``get_last_ordered_products_to_print`` (see : https://codecov.io/gh/OCA/product-attribute/compare/d66fc5d8bbb9e08d4ccf50f713107953cd648d3c...bd8eb83a2b3509cf7ffc7a44d97ba0114b817bec/diff)

File: product_pricelist_direct_print/readme/CONFIGURE.rst [hunk 1]
```diff
@@ -0,0 +1,7 @@
+To configure this module, you need to:
+
+#. Go to *Apps* and install *Sales Management*
+#. Go to *Sales > Configuration > Settings*
+#. Scroll to *Pricing*
+#. Set *Multiple Sales Prices per Product* on an either
```
Comment (Tardo) [inline product_pricelist_direct_print/readme/CONFIGURE.rst]:
```suggestion
#. Enable 'Pricelists'
```

File: product_pricelist_direct_print/readme/CONFIGURE.rst [hunk 2]
```diff
@@ -0,0 +1,7 @@
...
+   *Multiple prices per product* or *Prices computed from formulas*
```
Comment (Tardo) [inline product_pricelist_direct_print/readme/CONFIGURE.rst]:
```suggestion
```

Comment (Tardo) [reply product_pricelist_direct_print/readme/CONFIGURE.rst]: Can be good add a warning that the group 'Advanced Pricelists' is needed.

File: product_pricelist_direct_print/readme/USAGE.rst [hunk 3]
```diff
@@ -0,0 +1,15 @@
+To use this module, you have several options:
+
+#. Go to *Sales > Catalog > Print Price List*
```
Comment (Tardo) [inline product_pricelist_direct_print/readme/USAGE.rst]:
```suggestion
#. Go to *Sales > Products > Print Price List*
```

File: product_pricelist_direct_print/readme/USAGE.rst [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+
+#. Go to *Sales > Catalog > Products*
```
Comment (Tardo) [inline product_pricelist_direct_print/readme/USAGE.rst]:
```suggestion
#. Go to *Sales > Products > Products*
```

File: product_pricelist_direct_print/readme/USAGE.rst [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+    * Select products in list view
+    * Press *Print > Price List*
+
+#. Go to *Sales > Catalog > Product Variants*
```
Comment (Tardo) [inline product_pricelist_direct_print/readme/USAGE.rst]:
```suggestion
#. Go to *Sales > Products > Product Variants*
```

File: product_pricelist_direct_print/readme/USAGE.rst [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+    * Select products in list view
+    * Press *Print > Price List*
+
+#. Go to *Sales > Orders > Customers**
```
Comment (Tardo) [inline product_pricelist_direct_print/readme/USAGE.rst]:
```suggestion
#. Go to *Sales > Products > Customers**
```

File: product_pricelist_direct_print/readme/USAGE.rst [hunk 7]
```diff
@@ -0,0 +1,15 @@
+To use this module, you have several options:
+
+#. Go to *Sales > Catalog > Print Price List*
+
+#. Go to *Sales > Catalog > Products*
+    * Select products in list view
+    * Press *Print > Price List*
```
Comment (Tardo) [inline product_pricelist_direct_print/readme/USAGE.rst]: I can't see this option

**Review by Tardo — CHANGES_REQUESTED** [review]

**Comment by sergio-teruel** [comment]
@Tardo @legalsylvain Changes done!!

File: product_pricelist_direct_print/readme/CONFIGURE.rst [hunk 8]
```diff
@@ -0,0 +1,6 @@
+To configure this module, you need to:
+
+#. Go to *Apps* and install *Sales Management*
+#. Go to *Sales > Configuration > Settings*
+#. Scroll to *Pricing*
+#. Enable *Pricelists* > *Advanced price rules*.
```
Comment (Tardo) [inline product_pricelist_direct_print/readme/CONFIGURE.rst]:
```suggestion
#. Enable *Pricelists*

Users will need *Advanced Pricelists* to access the menus:

#. Enable developer mode
#. Go to *Settings > Users & Companies > Users*
#. Create or Edit a record
#. Scroll to *Technical Settings*
#. Enable *Advanced Pricelists*
```

File: product_pricelist_direct_print/readme/USAGE.rst (hunk 7 above)
Comment (Tardo) [reply product_pricelist_direct_print/readme/USAGE.rst]: This is shown in "Action" menu... not in "Print" :/

File: product_pricelist_direct_print/readme/USAGE.rst (hunk 6 above)
Comment (Tardo) [reply product_pricelist_direct_print/readme/USAGE.rst]: See this

**Review by Tardo — CHANGES_REQUESTED** [review]

**Comment by sergio-teruel** [comment]
@Tardo Changes done!

**Review by carlosdauden — APPROVED** [review]
Code and functional review LGTM
Fix precommit error

**Comment by simahawk** [comment]
@Tardo ping
@sergio-teruel the build needs some love ;)

File: product_pricelist_direct_print/wizards/product_pricelist_print.py [hunk 9]
```diff
@@ -0,0 +1,217 @@
...
+            self.partner_count = len(record.partner_ids)
+
+    @api.onchange("partner_ids")
+    def _onchange_partner_ids(self):
+        if not self.partner_count:
+            self.last_ordered_products = False
+
+    @api.model
+    def default_get(self, fields):
+        res = super(ProductPricelistPrint, self).default_get(fields)
```
Comment (Tardo) [inline product_pricelist_direct_print/wizards/product_pricelist_print.py]:
```suggestion
        res = super().default_get(fields)
```

File: product_pricelist_direct_print/wizards/product_pricelist_print.py [hunk 10]
```diff
@@ -0,0 +1,217 @@
...
+        else:
+            if self.partner_count == 1:
+                partner = self.partner_ids[0]
+                self.write(
+                    {
+                        "partner_id": partner.id,
+                        "pricelist_id": partner.property_product_pricelist.id,
+                    }
+                )
+            return self.message_composer_action()
```
Comment (Tardo) [inline product_pricelist_direct_print/wizards/product_pricelist_print.py]: Can reduce complexity...
```suggestion
            self.send_batch()
            return
        if self.partner_count == 1:
            partner = self.partner_ids[0]
            self.write(
                {
                    "partner_id": partner.id,
                    "pricelist_id": partner.property_product_pricelist.id,
                }
            )
        return self.message_composer_action()
```

**Review by Tardo — APPROVED** [review]
:+1: LGTM, only non-blocking comments

**Comment by sergio-teruel** [comment]
@Tardo changes done!!

File: product_pricelist_direct_print/wizards/product_pricelist_print.py [hunk 11]
```diff
@@ -0,0 +1,217 @@
...
+                    "product_tmpl_id.product_variant_ids"
+                )
+                product_ids |= product_ids.search(
+                    [
+                        ("sale_ok", "=", True),
+                        ("categ_id", "in", category_items.mapped("categ_id").ids),
+                    ]
+                )
+                res["product_ids"] = [(6, 0, product_ids.ids)]
+            # Convert al pricelist items to their affected templates
```
Comment (emagdalenaC2i) [inline product_pricelist_direct_print/wizards/product_pricelist_print.py]:
```suggestion
            # Convert all pricelist items to their affected templates
```

File: product_pricelist_direct_print/wizards/product_pricelist_print.py [hunk 12]
```diff
@@ -0,0 +1,217 @@
...
+            if len(items.mapped("pricelist_id")) == 1:
+                res["pricelist_id"] = items[0].pricelist_id.id
+            product_items = items.filtered(
+                lambda x: x.applied_on == "0_product_variant"
+            )
+            template_items = items.filtered(lambda x: x.applied_on == "1_product")
+            category_items = items.filtered(
+                lambda x: x.applied_on == "2_product_category"
+            )
+            # Convert al pricelist items to their affected variants
```
Comment (emagdalenaC2i) [inline product_pricelist_direct_print/wizards/product_pricelist_print.py]:
```suggestion
            # Convert all pricelist items to their affected variants
```

File: product_pricelist_direct_print/wizards/product_pricelist_print.py [hunk 13]
```diff
@@ -0,0 +1,217 @@
...
+        ctx = {
+            "default_composition_mode": "comment",
+            "default_res_id": self.id,
+            "default_model": "product.pricelist.print",
+            "default_use_template": bool(template_id),
+            "default_template_id": template_id,
+        }
+        return {
+            "type": "ir.actions.act_window",
+            "view_type": "form",
```
Comment (emagdalenaC2i) [inline product_pricelist_direct_print/wizards/product_pricelist_print.py]: I think you can delete this line, rigth?

**Review by emagdalenaC2i — APPROVED** [review]
Code LGTM, just small comments

**Comment by sergio-teruel** [comment]
Ok.. @emagdalenaC2i changes done!!

**Comment by pedrobaeza** [comment]
Isn't this missing the latest changes in previous versions?

File: product_pricelist_direct_print/views/report_product_pricelist.xml [hunk 14]
```diff
@@ -0,0 +1,136 @@
...
+<!-- Copyright 2017 Carlos Dauden <carlos.dauden@tecnativa.com>
+     License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl.html). -->
+<odoo>
+
+<template id="report_product_pricelist_document">
+    <t t-call="web.html_container">
+        <t t-call="web.external_layout">
+            <t t-set="pricelist" t-value="o.get_pricelist_to_print()"/>
+            <div class="page">
+                <h2>Price List</h2>
```
Comment (emagdalenaC2i) [inline product_pricelist_direct_print/views/report_product_pricelist.xml]: @pedrobaeza All changes in #581 are implemented except this line that was deleted.
But, I am not sure that it should be deleted

**Comment by pedrobaeza** [comment]
And what about #499 and #501?

**Comment by emagdalenaC2i** [comment]
No, those PRs have not yet migrated to v12. As I mentioned here https://github.com/OCA/product-attribute/pull/581#issuecomment-592173058

**Comment by pedrobaeza** [comment]
That's why I ask to migrate it here too before merging.

**Comment by simahawk** [comment]
@sergio-teruel are you going to attend https://github.com/OCA/product-attribute/pull/562#issuecomment-595398358 then?

**Comment by sergio-teruel** [comment]
@simahawk Last comments attended and added lost PR's

**Comment by sergio-teruel** [comment]
Rebase done!!

**Comment by simahawk** [comment]
@pedrobaeza good for you?

**Comment by ventor-dev** [comment]
@pedrobaeza , is it ok to be merged?

**Pull request merged**